File: day-03/part-2.ts
import { loadInput, sum } from "../utils";


const splitGroups = (rows: string[]) => {
  const groups = [];

  while (rows.length) {
    const group = rows.splice(0, 3);
    groups.push(group);
  }

  return groups;
};

const getCommons = (a: string, b: string, c: string) => {
  const ab = a.split("").filter(f => b.includes(f));
  const commons = ab.filter(f => c.includes(f));
  return commons;
};

const priority = (letter: string) => {
  const ascii = letter.charCodeAt(0);
  const isLowerCase = ascii >= 97;
  return isLowerCase ? ascii - 96 : ascii - 64 + 26;
};

const input = loadInput("input");
const rows = input.split("\n");

const groups = splitGroups(rows);
const letters = groups.flatMap(([a, b, c]) => [...new Set(getCommons(a, b, c))]);

const prios = letters.map(priority);

console.log(sum(prios));

File: kotlin/src/main/kotlin/day03/part-2.kt
package day03.part2

import loadInput

fun main() {
    val input = loadInput("./src/main/kotlin/day03/input")
    val rucksacks = input.split("\n")
    val groups = rucksacks.chunked(3)

    val intersections = groups.flatMap { it[0].toSet().intersect(it[1].toSet()).intersect(it[2].toSet()) }
    val prios = intersections.map { toPriority(it) }

    println(prios.sum())
}

fun toPriority(s: Char) = s.code - if (s.isLowerCase()) 96 else 64 - 26
